### ISSUE TITLE
Filtering query and using limit with a ManyToMany relation does not return the expected number of results

### ISSUE THREAD

**Issue by service-paradis** [mikro-orm/mikro-orm#1982]
**Describe the bug**

If there is enough results, using `limit` with a given number should return the given amount of results. For example, if a query would return 10 results, using `{ limit: 5 }` should return 5 results. Right now, if there is a `ManyToMany` relation, the number of returned results can be below 5.

**To Reproduce**

I'm using MariaBD. Here's the minimal entities needed to reproduce the issue with sample data.

`Resource.entity.ts`
```javascript
@Entity()
class Resource {
  @PrimaryKey()
  resourceId!: number;

  @ManyToMany({
    owner: true,
    entity: () => Grade,
    joinColumn: 'resource_id',
    inverseJoinColumn: 'grade_id',
  })
  grades!: Grade[];
}
```

`Grade.entity.ts`
```javascript
@Entity()
class Grade {
  @PrimaryKey()
  gradeId!: number;
}
```

Data for table `resources`:
```json
[
  {
    "resourceId": 1
  },
  {
   "resourceId": 2
  },
  {
    "resourceId": 3
  },
  {
    "resourceId": 4
  },
  {
    "resourceId": 5
  },
  {
    "resourceId": 6
  },
  {
    "resourceId": 7
  }
]
```

Data for table `grades`:
```json
[
  {
    "gradeId": 1
  },
  {
    "gradeId": 2
  },
  {
    "gradeId": 3
  }
]
```

Data for table `resources_grades`:
```json
[
  {
    "resourceId": 1,
    "gradeId": 5
  },
  {
    "resourceId": 2,
    "gradeId": 2
  },
  {
    "resourceId": 3,
    "gradeId": 1
  },
  {
    "resourceId": 3,
    "gradeId": 2
  },
  {
    "resourceId": 4,
    "gradeId": 3
  },
  {
    "resourceId": 5,
    "gradeId": 1
  },
  {
    "resourceId": 6,
    "gradeId": 2
  },
  {
    "resourceId": 7,
    "gradeId": 1
  },
  {
    "resourceId": 7,
    "gradeId": 2
  },
  {
    "resourceId": 7,
    "gradeId": 3
  },
]
```

Using `this.resourceRepository.find({ grades: [1, 2] })` will return 6 results (resource 1, 2, 3, 5, 6 and 7).
Using `this.resourceRepository.find({ grades: [1, 2] }, { limit: 5 })` will return 4 results (resource 1, 2, 3, 5).

**Expected behavior**

Using `this.resourceRepository.find({ grades: [1, 2] }, { limit: 5 })` I would expect 5 results (resource 1, 2, 3, 5, 6).

**Additional context**

I think I know what is going on. Using `this.resourceRepository.find({ grades: [1, 2] }, { limit: 5 })` will create the following query:
```sql
select `e0`.* from `resources` as `e0` left join `resources_grades` as `e1` on `e0`.`resource_id` = `e1`.`resource_id` where `e1`.`grade_id` in (1, 2) limit 5
```

If we run this in a SQL editor, it will return the following rows:
| resource_id |
| - |
| 1 |
| 2 |
| 3 |
| 3 |
| 5 |

I would expect the following query instead (with `DISTINCT`):
```sql
select DISTINCT `e0`.* from `resources` as `e0` left join `resources_grades` as `e1` on `e0`.`resource_id` = `e1`.`resource_id` where `e1`.`grade_id` in (1, 2) limit 5
```

Which gives the following rows:
| resource_id |
| - |
| 1 |
| 2 |
| 3 |
| 5 |
| 6 |

**Versions**

| Dependency | Version |
| - | - |
| node | 14.16.0 |
| typescript | 4.2.4 |
| mikro-orm | 4.5.6 |
| nestjs | 7.6.18 |

**Comment by B4nan** [comment]
This works as expected, `FindOptions.limit` simply controls the `limit` clause of SQL queries, you should not use it when fetch joining to many relations. To get around it you can use `QueryFlag.PAGINATE`, which will use subqueries to first get the PKs and then apply the limit on the parent query.

```ts
this.resourceRepository.find({ grades: [1, 2] }, { limit: 5, flags: [QueryFlag.PAGINATE] })
```

Btw your proposed fix would not really help, distict is applied to the whole tuple you are selecting, not just to the first key. In other words, distinct would help only if you select just the one column, not usable in realworld scenarios. 

Also note your entity definition is wrong, you need to use `Collection` wrapper, so instead of `grades!: Grade[];` it has to be `grades = new Collection<Grade>(this);`.

**Comment by service-paradis** [comment]
Thank you for your answer, that was fast :)

> This works as expected, FindOptions.limit simply controls the limit clause of SQL queries, you should not use it when fetch joining to many relations. To get around it you can use QueryFlag.PAGINATE, which will use subqueries to first get the PKs and then apply the limit on the parent query.

I totally missed `[QueryFlag.PAGINATE]`, sorry. Looks like it works as expected 🎉 

> Btw your proposed fix would not really help, distict is applied to the whole tuple you are selecting, not just to the first key. In other words, distinct would help only if you select just the one column, not usable in realworld scenarios.

Sorry for the confusion, I only wanted to let you know how to get the expected results from the data sample. It was not mean to be a fix.

> Also note your entity definition is wrong, you need to use Collection wrapper, so instead of grades!: Grade[]; it has to be grades = new Collection<Grade>(this);.

Totally right, sorry! I tried to create a minimal repro, but wrote it too fast.

Thank you very much for your work on this library by the way! I'm migrating to it and hope to use it in production this summer!

**Comment by B4nan** [comment]
> Sorry for the confusion, I only wanted to let you know how to get the expected results from the data sample. It was not mean to be a fix.

Yeah no worries, just wanted to point out why it won't work, as I already tried :]

Looks like it is not documented, will try to add it somewhere to the docs. I also wanted to make this work automatically, but afaik it is problematic in mysql 8 (test suite runs on 5.7 and this was one of the tests that failed when I tried to run it on 8.0). Will need to review that before v5 is out together with the docs, there are many areas that should be updated.
